### PR TITLE
Rewrite prediction splitting adding support for regex separators

### PR DESCRIPTION
**Pull request by mawelborn**
Simplifies the implementation of prediction splitting and adds support for regular expression separators.

### CODE REVIEW

File: indico_toolkit/association/splitting.py [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+    """
+    Split a single `prediction` into mutiple predictions along a `separator` regex
+    and yield new predictions having updated text and spans.
+    """
+    texts_and_delimiters = re.split(f"({separator})", prediction["text"])
+    texts = texts_and_delimiters[::2]
+    delimiters = texts_and_delimiters[1::2]
+    text_and_delimiter_pairs = zip_longest(texts, delimiters)
+    start = prediction["start"]
+    end = 0
```
Comment (Scott771) [inline indico_toolkit/association/splitting.py:19]: don't believe you need to define end outside of the loop

File: indico_toolkit/association/splitting.py [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+
+def split_prediction(
+    prediction: Dict[str, Any], separator: str = r"\s*\n\s*"
+) -> Iterator[Dict[str, Any]]:
+    """
+    Split a single `prediction` into mutiple predictions along a `separator` regex
+    and yield new predictions having updated text and spans.
+    """
+    texts_and_delimiters = re.split(f"({separator})", prediction["text"])
+    texts = texts_and_delimiters[::2]
```
Comment (Scott771) [inline indico_toolkit/association/splitting.py:15]: I'd add comment in here w/ the rationale for staggering the texts/delimiters like this w/ zip_longest

File: indico_toolkit/association/splitting.py [hunk 3]
```diff
@@ -0,0 +1,32 @@
+import re
+from copy import deepcopy
+from itertools import zip_longest
+from typing import Any, Dict, Iterator
+
+
+def split_prediction(
+    prediction: Dict[str, Any], separator: str = r"\s*\n\s*"
```
Comment (Scott771) [inline indico_toolkit/association/splitting.py:8]: I'd define what the default separator is going to split on in the docstring

File: indico_toolkit/association/splitting.py (hunk 1 above)
Comment (mawelborn) [reply indico_toolkit/association/splitting.py]: Good catch! Removed.

File: indico_toolkit/association/splitting.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (mawelborn) [reply indico_toolkit/association/splitting.py:25]: I added an explanation for the pairing logic with examples at each step.

File: indico_toolkit/association/splitting.py (hunk 3 above)
Comment (mawelborn) [reply indico_toolkit/association/splitting.py:8]: Added to the docstring that the default behavior splits on newlines and also captures any adjacent whitespace.

**Review by Scott771 — APPROVED** [review]

**Pull request merged**